File: source/WeddingManager.Web/js/services/customerInteractionService.js
(function(){
    angular.module('weddingManager')
    .service('customerInteractionService', function(customerInteractionFactory, appStateService){
        var service = {};
        var _customerInteractions = [];
        var _customerInteractionTypes = [];
        
        service.createCustomerInteraction = function(customerId, onSuccess, onError){
            var customerInteraction = {
                Date : new Date(),
                InteractionType : 1,
                Notes : ""
            };
            customerInteractionFactory.createCustomerInteraction(customerId, customerInteraction, function(customerInteractionId){
                customerInteraction.Id = customerInteractionId;
                onSuccess(customerInteraction);
            }, onError);
        }
                
        service.refreshCustomerInteractions = function(customerId, onSuccess, onError){
            customerInteractionFactory.retrieveCustomerInteractions(customerId, function(customerInteractions){
                _customerInteractions = customerInteractions;
                onSuccess(_customerInteractions);                
            }, onError);
        }
                
        service.updateCustomerInteraction = function(customerInteraction, onSuccess, onError){
            customerInteractionFactory.updateCustomerInteraction(customerInteraction, function(){
                onSuccess();
            }, onError);
        }
        
        service.deleteCustomerInteraction = function(customerInteractionId, onSuccess, onError){
            customerInteractionFactory.deleteCustomerInteraction(customerInteractionId, function(){
                appStateService.setCustomerInteraction(null);
                onSuccess();
            }, onError);
        }
        
        service.getCustomerInteractionTypes = function(onSuccess, onError){
            if(_customerInteractionTypes.length == 0){
                customerInteractionFactory.retrieveCustomerInteractionTypes(function(customerInteractionTypes){
                    _customerInteractionTypes = customerInteractionTypes;
                    onSuccess(_customerInteractionTypes);                
                }, onError);
            } else {
                onSuccess(_customerInteractionTypes);
            }
        }
        
        return service;
    });
})();

File: source/WeddingManager.Services/CustomerInteractionService.cs
using System.Collections.Generic;
using WeddingManager.Core.Data;
using WeddingManager.Core.Repositories;
using WeddingManager.Core.Services;

namespace WeddingManager.Services
{
    public class CustomerInteractionService : ICustomerInteractionService
    {
        public CustomerInteractionService(ICustomerInteractionRepository customerInteractionRepository)
        {
            _customerInteractionRepository = customerInteractionRepository;
        }

        public int CreateCustomerInteraction(int customerId, CustomerInteraction customerInteraction)
        {
            return _customerInteractionRepository.CreateCustomerInteraction(customerId, customerInteraction);
        }

        public IEnumerable<CustomerInteraction> RetrieveCustomerInteractions(int customerId)
        {
            return _customerInteractionRepository.RetrieveCustomerInteractions(customerId);
        }

        public void UpdateCustomerInteraction(CustomerInteraction customerInteraction)
        {
            _customerInteractionRepository.UpdateCustomerInteraction(customerInteraction);
        }

        public void DeleteCustomerInteraction(int customerInteractionId)
        {
            _customerInteractionRepository.DeleteCustomerInteraction(customerInteractionId);
        }

        public IDictionary<int, string> RetrieveCustomerInteractionTypes()
        {
            return _customerInteractionRepository.RetrieveCustomerInteractionTypes();
        }

        private readonly ICustomerInteractionRepository _customerInteractionRepository;
    }
}